# Regressão 01 - tarefa 02 - transformações em X e Y

In [36]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

import statsmodels.api as sm

import warnings
#Algumas células estavam retornando com um aviso, por isso o uso dessa biblioteca
warnings.filterwarnings('ignore', message='omni_normtest is not valid with less than 8 observations')


### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [9]:
df = pd.read_csv('dataFrame.csv')
df

Unnamed: 0  total_bill   tip     sex smoker  day    time  size  tips_pct  \
0           0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204   
1           1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244   
2           2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886   
3           3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494   
4           4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069   

   net_bill  
0     15.98  
1      8.68  
2     17.51  
3     20.37  
4     20.98

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [20]:
df['log_net_bill'] = np.log(df['net_bill'])

In [38]:
y = df['tips_pct'] #Target
X = df['log_net_bill'] 

X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:               tips_pct   R-squared:                       0.021
Model:                            OLS   Adj. R-squared:                 -0.305
Method:                 Least Squares   F-statistic:                   0.06586
Date:                Fri, 03 Jan 2025   Prob (F-statistic):              0.814
Time:                        13:49:02   Log-Likelihood:                 8.0271
No. Observations:                   5   AIC:                            -12.05
Df Residuals:                       3   BIC:                            -12.84
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.2200      0.244      0.901   

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [40]:
df['net_bill_squared'] = df['net_bill'] ** 2

In [42]:
y = df['tips_pct']
X = df[['net_bill', 'net_bill_squared']]

In [44]:
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:               tips_pct   R-squared:                       0.310
Model:                            OLS   Adj. R-squared:                 -0.380
Method:                 Least Squares   F-statistic:                    0.4489
Date:                Fri, 03 Jan 2025   Prob (F-statistic):              0.690
Time:                        13:49:06   Log-Likelihood:                 8.8998
No. Observations:                   5   AIC:                            -11.80
Df Residuals:                       2   BIC:                            -12.97
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const                0.5077      0.375  

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [46]:
df['log_tip_pct'] = np.log(df['tips_pct'])

y_log = df['log_tip_pct']
X = df['net_bill']

In [48]:
X = sm.add_constant(X)

model_log = sm.OLS(y_log, X).fit()

In [50]:
print("Resumo do modelo log(tip_pct) ~ net_bill")
print(model_log.summary())

Resumo do modelo log(tip_pct) ~ net_bill
                            OLS Regression Results                            
Dep. Variable:            log_tip_pct   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.333
Method:                 Least Squares   F-statistic:                  0.001520
Date:                Fri, 03 Jan 2025   Prob (F-statistic):              0.971
Time:                        13:51:21   Log-Likelihood:                -2.8365
No. Observations:                   5   AIC:                             9.673
Df Residuals:                       3   BIC:                             8.892
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      

In [52]:
r2_log = model_log.rsquared
print(f"R² do modelo com log(tip_pct): {r2_log:.4f}")

R² do modelo com log(tip_pct): 0.0005


In [54]:
y = df['tips_pct']
model_original = sm.OLS(y, X).fit()

# R² do modelo original
r2_original = model_original.rsquared
print(f"R² do modelo original (tip_pct) ~ net_bill: {r2_original:.4f}")

R² do modelo original (tip_pct) ~ net_bill: 0.0066


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [56]:
df['log_tip_pct'] = np.log(df['tips_pct'])
df['log_net_bill'] = np.log(df['net_bill'])

y_log = df['log_tip_pct']
X = df['log_net_bill']

X = sm.add_constant(X)

model_log = sm.OLS(y_log, X).fit()
print("Resumo do modelo log(tip_pct) ~ log(net_bill)")
print(model_log.summary())

Resumo do modelo log(tip_pct) ~ log(net_bill)
                            OLS Regression Results                            
Dep. Variable:            log_tip_pct   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                 -0.321
Method:                 Least Squares   F-statistic:                   0.02759
Date:                Fri, 03 Jan 2025   Prob (F-statistic):              0.879
Time:                        13:53:31   Log-Likelihood:                -2.8149
No. Observations:                   5   AIC:                             9.630
Df Residuals:                       3   BIC:                             8.849
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
co

In [58]:
r2_log = model_log.rsquared
print(f"R² do modelo com log(tip_pct) ~ log(net_bill): {r2_log:.4f}")

y = df['tips_pct']
model_original = sm.OLS(y, X).fit()

# R² do modelo original
r2_original = model_original.rsquared
print(f"R² do modelo original (tip_pct) ~ log(net_bill): {r2_original:.4f}")

R² do modelo com log(tip_pct) ~ log(net_bill): 0.0091
R² do modelo original (tip_pct) ~ log(net_bill): 0.0215


###  6. qual modelo resultou no melhor ajuste?

### R:
O modelo que gerou o melhor ajuste foi o modelo original (tip_pct) ~ log(net_bill): 0.0215